src/transformStateWithClones.js

```
@@ -1,13 +1,33 @@
-'use strict';
-
 /**
  * @param {Object} state
  * @param {Object[]} actions
- *
- * @return {Object[]}
+ * @returns {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'clear':
+        currentState = {};
+        break;
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+      case 'removeProperties':
+        currentState = { ...currentState };
+
+        for (const key of action.keysToRemove || []) {
+          delete currentState[key];
+        }
+        break;
+    }
```

This violates checklist item #3: "switch/case should always have default case for error handling." The `switch (action.type)` (lines 11–25) includes cases for 'clear', 'addProperties', and 'removeProperties' but no `default` branch. Add a `default` case to handle unknown action types (for example, throw a descriptive `Error`) so invalid actions are not silently ignored. See checklist requirement for this rule  and the current implementation for context .